### PR TITLE
MWS Gaia DR2->EDR3

### PR DESCRIPTION
**Pull request by apcooper**
Changes in response to switching MWS targeting to EDR3.

Including:
- Update AEN threshold to 2 from 3
- Lower error floor in RED parallax cut to 0.5 from 1 mas

Some final checks are in progress in MWS to validate these cuts on the EDR3 sweeps. Timeframe is 1-2 days.

Will need to update FAINT cuts in #719 once merged.

### CODE REVIEW

**Comment by geordie666** [comment]
Note that we only update parallax and proper motion to EDR3. The AEN still remains Gaia DR2. I think we could also update the AEN to EDR3. But, we don't currently. Updating the AEN would require an additional change to this function:

https://github.com/desihub/desitarget/blob/8e3b3d6d23cbbfea51693ee2c7072473415901bf/py/desitarget/gaiamatch.py#L195

(at least, I think that's the meat of where I decided to make the DR2->EDR3 substitutions in the sweeps).

As a rule, we should only allows updates to parameters to EDR3 if they are only used by the MWS (and standard star) selections.

**Comment by geordie666** [comment]
Also, I was hedging my bets on whether we would make the DR2->EDR3 updates. If we're definitely going to do that, we might want to switch the sense of this:

https://github.com/desihub/desitarget/blob/8e3b3d6d23cbbfea51693ee2c7072473415901bf/bin/select_targets#L43-L44

So that the default is `gaiasub` (for the Main Survey) and `nogaiasub` would need to be passed to turn that behavior _off_.

**Comment by apcooper** [comment]
Thanks for the comments, @geordie666.

> Note that we only update parallax and proper motion to EDR3. The AEN still remains Gaia DR2. I think we could also update the AEN to EDR3. But, we don't currently. 

I feel we should, but @segasai's opinion would be welcome on how important/useful this is.

> As a rule, we should only allows updates to parameters to EDR3 if they are only used by the MWS (and standard star) selections.

I can't see it being used anywhere except these cases. BGS have a different star/galaxy separation.

**Comment by apcooper** [comment]
Also, just noticed that AEN is used in the WD selection, so we should double check with Boris that this change is not significant there.

**Comment by apcooper** [comment]
@segasai and @geordie666 might also want to consider whether we need to update this function with EDR3 cuts if AEN is updated to EDR3:

https://github.com/desihub/desitarget/blob/8e3b3d6d23cbbfea51693ee2c7072473415901bf/py/desitarget/gaiamatch.py#L167-L168

**Comment by geordie666** [comment]
If we do update the `gaia_psflike` function to have different cuts for `edr3`, we will need to retain backwards compatibility for any calls in other codebases that use it. For example, with a `dr="dr2"` kwarg that, if passed as `edr3` uses the different set of cuts.

**Comment by BGaensicke** [comment]
We were planning to use for the WD selection phot_bp_rp_excess_factor and astrometric_sigma5d_max: 

```
    # ADM As of DR7, photbprpexcessfactor and astrometricsigma5dmax are not in the
    # ADM imaging catalogs. Until they are, ignore these cuts
    if photbprpexcessfactor is not None:
        # ADM remove problem objects, which often have bad astrometry
        mws &= photbprpexcessfactor < 1.7 + 0.06*br*br

    if astrometricsigma5dmax is not None:
        # ADM Reject white dwarfs that have really poor astrometry while
        # ADM retaining white dwarfs that only have relatively poor astrometry
        mws &= ((astrometricsigma5dmax < 1.5) |
                ((astrometricexcessnoise < 1.) & (parallaxovererror > 4.) & (pm > 10.)))
```

but I'm not sure if those parameters were available in the DR2 sweeps, in other words, we may have run the selection without those? If that's the case, then adding them now would probably cause much pain and we should just drop those statements.

**Comment by geordie666** [comment]
@BGaensicke: We definitely use those. The "primary" targets are run on the Legacy Surveys sweeps files, not the Gaia files. Those columns are [included in the Legacy Surveys sweeps files](https://www.legacysurvey.org/dr9/files/#sweep-catalogs-region-sweep).

**Comment by BGaensicke** [comment]
If we do update AEN to EDR3, we should also update PHOT_BP_RP_EXCESS_FACTOR and ASTROMETRIC_SIGMA5D_MAX, so that the selection runs on a consistent set of Gaia data. The WD selection algorithm would remain the same, just using the updated data.

**Comment by geordie666** [comment]
`PHOT_BP_RP_EXCESS_FACTOR` is used by other selections (most notably the standard stars). So, if it's critical to have consistency in the WD selection then we _shouldn't_ update AEN either.

**Comment by BGaensicke** [comment]
OK, I think I see: the Gaia photometry will remain DR2, right? 

I think that the biggest impact that EDR3 has is in the parallaxes and proper motions, updating  PHOT_BP_RP_EXCESS_FACTOR and ASTROMETRIC_SIGMA5D_MAX will only result in subtle changes, so if that causes headaches with other target classes, sticking to DR2 for those two parameters is fine.

**Comment by geordie666** [comment]
Thanks @BGaensicke. And, yes, the Gaia photometry will all remain DR2.

**Comment by segasai** [comment]
I looked at the code, and I think we must use the astrometric_params_solved from EDR3 and we need replace the astrometric_nparams==31 by >=31 (because it can have values of 3, 31, 95).
Regarding AEN, we are good with AEN<2 new cut. and it's unclear if we need to change gaia_psflike() as I don't see it being called by any target selection code.
I'll send info on new plx,pm selection

**Comment by geordie666** [comment]
Thanks for your checks Sergey. It would be great if we don't update `gaia_psflike()`. It _is_ called by functions outside of the main target selection code (notably to select GFAs).

Great catch on `astrometric_params_solved`. I agree that will need to be updated to `>=31` everywhere it's used.

**Comment by segasai** [comment]
Sure. I've created a different  branch ( I didn't know if I should tinker with this one or not)
https://github.com/desihub/desitarget/commits/mws_edr3x
https://github.com/desihub/desitarget/commit/b05985d073478fd25b2b9c76eae6a92be0ddca8a
with the 
* Andrew's AEN change, 
* the nparams fix.
* Update to the MW plx cut. 
I intend to apply a final tweak to the MWS proper motion cut as well tomorrow (or keep it as it is)

**Comment by geordie666** [comment]
@segasai: It's Andrew's branch so it's up to him whether he wants to let you tinker with it directly. I'm fine with you doing that if @apcooper signs off on direct tinkering, though.

**Comment by apcooper** [comment]
@segasai, I'm teaching a class now (!)so I'll look at this again later tonight but please go ahead and tinker with this branch!

**Comment by apcooper** [comment]
OK, so @segasai, if you can please merge your branch with this one after you've made the decision about the proper motion cut, then I'll take care of any loose ends like `gaia_psflike`, update the target selection page on the wiki and do the final merge.

Regarding `gaia_psflike`, it seems to be used in selecting standards and setting the `TYPE` column for the gfa targets. The question is whether the cut of 3 in `gaia_psflike` would be harmful in those cases if we apply it to the EDR3 AEN values. I'm leaning towards Adam's idea of having a switch and changing 3 to 2 in the EDR3 case, rather than using `AEN < 3` for all Gaia releases. 

However, having read the above discussion I'm still unclear what if anything changes in the WD selection and which columns are EDR3 and which are DR2. I *think* `PHOT_BP_RP_EXCESS_FACTOR`, `ASTROMETRIC_SIGMA5D_MAX` and the photometry are staying as DR2, but AEN will be EDR3.

**Comment by BGaensicke** [comment]
>
> However, having read the above discussion I'm still unclear what if
> anything changes in the WD selection and which columns are EDR3 and which
> are DR2. I *think* PHOT_BP_RP_EXCESS_FACTOR, ASTROMETRIC_SIGMA5D_MAX and
> the photometry are staying as DR2, but AEN will be EDR3.
>
yes, that was the conclusion.

**Comment by segasai** [comment]
Sorry, it took me a while. I've updated the mws_edr3x branch with the new proper motion cut (as in the desi-mw email) and ensured that the NPARAMS are substituted as well from EDR3  
I can't merge my commits unfortunately here, because this branch does not include the commits from 
#719 while my branch was split after #719 landed. So either this branch need to be rebased, or we'll just work with mws_edr3x.

**Comment by apcooper** [comment]
OK I have a bit of time now so I'll rebase this branch and merge yours, Sergey.

**Comment by apcooper** [comment]
I've now merged @segasai 's branch with mine. 

I've also added the dr2/edr3 switch in `gaia_psflike` as discussed above, with default of dr2 for backwards compatibility. This is used in `isSTD_gaia`, which I've updated to use edr3 for the main survey only.

I'm not sure what to do about the one other place `gaia_psflike` is used, in the GFA targets:

https://github.com/desihub/desitarget/blob/05db4981cbd9ed918924dd7d2d97b91290ec5f8f/py/desitarget/gfa.py#L56-L57

called from

https://github.com/desihub/desitarget/blob/05db4981cbd9ed918924dd7d2d97b91290ec5f8f/py/desitarget/gfa.py#L244-L245

in
https://github.com/desihub/desitarget/blob/05db4981cbd9ed918924dd7d2d97b91290ec5f8f/py/desitarget/gfa.py#L161-L163

Either (1) the DR comes down through the stack above this, (2) it gets deduced somehow from the files that are being read here, or (3) we just ignore this one and leave it as DR2. @geordie666?

Finally, there are two other places AEN is used, perhaps we should double check. In `isSTD_gaia`:

https://github.com/desihub/desitarget/blob/e49c5b6640f8ab63447dc043975072746e38c733/py/desitarget/cuts.py#L675-L677

and in the WD selection we discussed above.

https://github.com/desihub/desitarget/blob/e49c5b6640f8ab63447dc043975072746e38c733/py/desitarget/cuts.py#L1250-L1254

These look fine, though they use a stricter cut than we have for the main sample.

If all that is ok, it seems like this branch is ready.

File: py/desitarget/cuts.py
Comment (geordie666) [inline py/desitarget/cuts.py:2842]: Revert all changes in `apply_cuts_gaia()`.

File: py/desitarget/cuts.py
Comment (geordie666) [inline py/desitarget/cuts.py:2807]: Revert all changes in `apply_cuts_gaia()`.

File: py/desitarget/cuts.py
Comment (geordie666) [inline py/desitarget/cuts.py:185]: Consider reverting all changes in `isGAIA_STD()`.

File: py/desitarget/cuts.py
Comment (geordie666) [inline py/desitarget/cuts.py:198]: Consider reverting all changes in `isGAIA_STD()`.

File: py/desitarget/cuts.py
Comment (geordie666) [inline py/desitarget/cuts.py:242]: Consider reverting all changes in `isGAIA_STD()`.

File: py/desitarget/gaiamatch.py
Comment (geordie666) [inline py/desitarget/gaiamatch.py:241]: I think this will fail? The parameter is called `EDR3_ASTROMETRIC_PARAMS_SOLVED` in the Gaia EDR3 sweeps files, so for this additional parameter you need to swap out `GAIA` for `EDR3` rather than just prepending `EDR3`?

File: py/desitarget/gaiamatch.py
Comment (geordie666) [inline py/desitarget/gaiamatch.py:240]: It looks like you made changes on AEN in the MWS main survey cuts. Are you absolutely sure you don't want to use the EDR3 version of astrometric_excess_noise, which would require transferring it in here?

**Review by geordie666 — CHANGES_REQUESTED** [review]
@apcooper, I've a few requested changes. The main important point is that `apply_cuts_gaia()` will still use Gaia DR2, regardless of any changes you've made so far, unless we change the [call to all_gaia_in_tiles()](https://github.com/desihub/desitarget/blob/38413c0c020dfc9cb969caac301a38d0ce50454c/py/desitarget/cuts.py#L2808-L2811) to acknowledge the existence of EDR3. Changing `all_gaia_in_tiles()` will have knock-on effects in lots of places in the code, though.

I suggest that we take the easy route; don't update `apply_cuts_gaia()` at all, and just acknowledge the fact that the `BACKUP` targets assigned by `apply_cuts_gaia()` will still run with Gaia DR2 parameters. That choice will also mean that your changes to `isGAIA_STD()` are obsolete, so you could remove those changes to (although I think they're harmless, so that's up to you).

File: py/desitarget/gaiamatch.py
Comment (apcooper) [reply py/desitarget/gaiamatch.py]: Good catch, thanks.

File: py/desitarget/gaiamatch.py
Comment (apcooper) [reply py/desitarget/gaiamatch.py:240]: Again, good catch. I've added AEN to this list and removed the GAIA prefix from APS.

File: py/desitarget/cuts.py
Comment (apcooper) [reply py/desitarget/cuts.py]: Reverted all changes.

File: py/desitarget/cuts.py
Comment (apcooper) [reply py/desitarget/cuts.py]: Reverted all changes here too.

File: py/desitarget/cuts.py
Comment (apcooper) [reply py/desitarget/cuts.py]: Reverted.

File: py/desitarget/cuts.py
Comment (apcooper) [reply py/desitarget/cuts.py]: Reverted.

File: py/desitarget/gaiamatch.py
Comment (geordie666) [reply py/desitarget/gaiamatch.py]: This is almost correct now, but not quite. You need to turn [these columns](https://github.com/desihub/desitarget/blob/5b2a280bb303eef17d49ec748e631047c4e8fb3a/py/desitarget/gaiamatch.py#L63-L66):
```
    ('GAIA_ASTROMETRIC_SIGMA5D_MAX', '>f4'), ('GAIA_ASTROMETRIC_PARAMS_SOLVED', '>i1'),
    ('PARALLAX', '>f4'), ('PARALLAX_IVAR', '>f4'),
    ('PMRA', '>f4'), ('PMRA_IVAR', '>f4'),
    ('PMDEC', '>f4'), ('PMDEC_IVAR', '>f4')
```
Into [these columns](https://github.com/desihub/desitarget/blob/5b2a280bb303eef17d49ec748e631047c4e8fb3a/py/desitarget/gaiamatch.py#L93-L98):
```
    ('EDR3_ASTROMETRIC_SIGMA5D_MAX', '>f4'), ('EDR3_ASTROMETRIC_PARAMS_SOLVED', '>i1'), 
    ('EDR3_PARALLAX', '>f4'), ('EDR3_PARALLAX_IVAR', '>f4'),
    ('EDR3_PMRA', '>f4'), ('EDR3_PMRA_IVAR', '>f4'),
    ('EDR3_PMDEC', '>f4'), ('EDR3_PMDEC_IVAR', '>f4')
```
So the `PMRA` and `PARALLAX` columns need an `EDR3` prepended but the `GAIA_ASTROMETRIC_SIGMA5D_MAX` and `GAIA_ASTROMETRIC_PARAMS_SOLVED` columns need `GAIA` switched to `EDR3`. So, some code logic like:

```
gaiacols = [col.replace("GAIA", "EDR3") if "GAIA" in col else "EDR3_{}".format(col) for col in cols]
```
should work.

Sorry this is so tricky. I inherited the Gaia DR2 data model from the Legacy Surveys catalogs.

**Comment by geordie666** [comment]
@apcooper: Sorry, but one more change requested (see the conversation that I "unresolved" above). Once you've corrected that, you can merge it if you like. I'll test the result and make any additional fixes (if I find that any are necessary) early tomorrow morning my time.

File: py/desitarget/gaiamatch.py
Comment (apcooper) [reply py/desitarget/gaiamatch.py]: OK, thanks. Sorry I'm clearly not following very well how the columns get copied. Will make those changes.

Also I thought the discussion with Boris above ended on GAIA_ASTROMETRIC_SIGMA5D_MAX being left as DR2. For consistency it seems better to me to change it, so I will follow your suggestion.

**Comment by apcooper** [comment]
@geordie666, many thanks indeed for your help with this. I made those changes but I also noted that we hadn't discussed GAIA_DUPLICATED_SOURCE -- it gets used in the MWS selection so I guess it should also be updated, although I hope that doesn't lead to some inconsistency for the selections that are still using dr2...

I'll hold off on the merge until either you or @segasai or both have had chance to comment on that (and perhaps GAIA_ASTROMETRIC_SIGMA5D_MAX, although I think that's a safer change).

**Comment by geordie666** [comment]
@apcooper:

- Regarding `GAIA_DUPLICATED_SOURCE`: I think it's fine to also update that. It's only every used in a `True`/`False` sense so I don't think it can ever be dangerous to substitute the EDR3 version into the mix. i.e. there won't be any changes to parameters/cuts.
- Regarding `GAIA_ASTROMETRIC_SIGMA5D_MAX`: It's only used (I think) by the WDs in the clause:
```
       mws &= ((astrometricsigma5dmax < 1.5) |
                ((astrometricexcessnoise < 1.) & (parallaxovererror > 4.) & (pm > 10.)))
```
I think it's fine to change it _if_ we don't need to change the 1.5 in the clause above, such that the EDR3 and DR2 parameter cuts remain identical for WDs.

**Comment by apcooper** [comment]
@BGaensicke: please can you confirm (one last time, sorry) that if we change `astrometricsigma5dmax` to the EDR3 value, we can keep the value of 1.5 in `(astrometricsigma5dmax < 1.5)` for the WD selection?

**Comment by apcooper** [comment]
@geordie666 if this branch is holding things up I think it's OK to merge now, I guess there might still be a chance to tweak astrometricsigma5dmax later...

**Comment by geordie666** [comment]
There might not be a chance to tweak anything else starting from about 6 hours from now. So, let's leave this open as long as we can. It's not holding anything up yet.

**Comment by BGaensicke** [comment]
@geordie666: The tests looked all good, no further changes from the WD side.

**Pull request merged**
